bundles/org.openhab.binding.dirigera/src/main/java/org/openhab/binding/dirigera/internal/handler/DirigeraHandler.java

```diff
@@ -0,0 +1,1056 @@
+            if (gatewayStorageObject != null) {
+                JSONObject gatewayStorageJson = new JSONObject(gatewayStorageObject.toString());
+                return gatewayStorageJson;
+            }
+        }
+        return new JSONObject();
+    }
+
+    /**
+     * Everything to handle pairing process
+     */
+
+    private String getCode(String challenge) {
+        try {
+            MultiMap<@Nullable String> baseParams = new MultiMap<>();
+            baseParams.put("audience", "homesmart.local");
+            baseParams.put("response_type", "code");
+            baseParams.put("code_challenge", challenge);
+            baseParams.put("code_challenge_method", "S256");
+
+            String url = String.format(OAUTH_URL, config.ipAddress);
+            Request codeRequest = httpClient.newRequest(url).param("audience", "homesmart.local")
+                    .param("response_type", "code").param("code_challenge", challenge)
+                    .param("code_challenge_method", "S256");
+
+            ContentResponse response = codeRequest.timeout(10, TimeUnit.SECONDS).send();
+            int responseStatus = response.getStatus();
+            if (responseStatus != 200) {
+                String reason = response.getReason();
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                        "@text/dirigera.gateway.status.comm-error" + " [\"" + responseStatus + " - " + reason + "\"]");
+                return "";
+            }
+            String responseString = response.getContentAsString();
+            JSONObject codeResponse = new JSONObject(responseString);
+            String code = codeResponse.getString("code");
+            return code;
+        } catch (InterruptedException | TimeoutException | ExecutionException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "@text/dirigera.gateway.status.comm-error" + " [\"" + e.getMessage() + "\"]");
+            return "";
+        }
+    }
+
+    private String getToken(String code, String codeVerifier) {
+        try {
+            MultiMap<@Nullable String> baseParams = new MultiMap<>();
+            baseParams.put("code", code);
+            baseParams.put("name", "openHAB");
+            baseParams.put("grant_type", "authorization_code");
+            baseParams.put("code_verifier", codeVerifier);
+
+            String url = String.format(TOKEN_URL, config.ipAddress);
+            String urlEncoded = UrlEncoded.encode(baseParams, StandardCharsets.UTF_8, false);
+            Request tokenRequest = httpClient.POST(url)
+                    .header(HttpHeader.CONTENT_TYPE, "application/x-www-form-urlencoded")
+                    .content(new StringContentProvider("application/x-www-form-urlencoded", urlEncoded,
+                            StandardCharsets.UTF_8))
+                    .followRedirects(true);
```

Comment:
Bettter to also set an explicit timeout

Reply:
Timeout defined one line after that but sure it can be moved to request creation.